### PR TITLE
Generic KV Client Instrumentation

### LINKED ISSUE #2484: Need a way to differentiate between KV metrics for different rings and dudupers in single binary mode.
Current its all in one histogram - `cortex_kv_request_duration_seconds_count`.  Need to add a name label or something.

### PR DESCRIPTION
**Pull request by jtlisi**
**What this PR does**:

- Add a name and backend type label to the prometheus metrics that wrap KV clients
- Remove dedicated instrumentation for the consul client in favor of a generic instrumentation for all clients

**Which issue(s) this PR fixes**:
Fixes #2484 

**Checklist**
- [x] Tests updated
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ring/kv/client.go [hunk 1]
```diff
@@ -97,19 +99,19 @@ type Client interface {
 
 // NewClient creates a new Client (consul, etcd or inmemory) based on the config,
 // encodes and decodes data for storage using the codec.
-func NewClient(cfg Config, codec codec.Codec) (Client, error) {
+func NewClient(name string, cfg Config, codec codec.Codec, reg prometheus.Registerer) (Client, error) {
```
Comment (pstibrany) [inline pkg/ring/kv/client.go:102]: I think passing `name` to components that is used just for metrics registration is an anti-pattern. Caller of this call can easily wrap `reg` into `prometheus.WrapRegistererWith(prometheus.Labels{"name": "some name here"}, reg)` if it wishes to do so. I would prefer to use this pattern instead. KV store client doesn't need a name for anything else, and it also doesn't need it if registrerer is `nil`.

File: pkg/ring/kv/client.go [hunk 2]
```diff
@@ -149,10 +151,10 @@ func createClient(name string, prefix string, cfg StoreConfig, codec codec.Codec
 		client = PrefixClient(client, prefix)
 	}
 
-	return metrics{client}, nil
+	return newMetricsClient(name, backend, client, reg), nil
```
Comment (pstibrany) [inline pkg/ring/kv/client.go:154]: If `reg` is nil (which is a case we should support and is typical for tests), we can skip creation of metrics client completely.

File: pkg/ring/lifecycler.go [hunk 3]
```diff
@@ -147,7 +147,8 @@ func NewLifecycler(cfg LifecyclerConfig, flushTransferer FlushTransferer, ringNa
 	}
 	port := GetInstancePort(cfg.Port, cfg.ListenPort)
 	codec := GetCodec()
-	store, err := kv.NewClient(cfg.RingConfig.KVStore, codec)
+	// Suffix all client names with "-lifecycler" to denote this kv client is used by the lifecycler
+	store, err := kv.NewClient(ringName+"-lifecycler", cfg.RingConfig.KVStore, codec, nil)
```
Comment (pstibrany) [inline pkg/ring/lifecycler.go:151]: This is a great opportunity to pass registrerer to lifecycler and move lifecycler metrics away from global variables.

File: pkg/distributor/ha_tracker_test.go [hunk 4]
```diff
@@ -135,7 +136,7 @@ func TestWatchPrefixAssignment(t *testing.T) {
 		UpdateTimeout:          time.Millisecond,
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Millisecond * 2,
-	})
+	}, prometheus.NewRegistry())
```
Comment (pstibrany) [inline pkg/distributor/ha_tracker_test.go:139]: `nil`, unless metrics are checked.

File: pkg/ring/kv/consul/client.go [hunk 5]
```diff
@@ -85,7 +84,7 @@ func NewClient(cfg Config, codec codec.Codec) (*Client, error) {
 		return nil, err
 	}
 	c := &Client{
-		kv:    consulMetrics{client.KV()},
```
Comment (pstibrany) [inline pkg/ring/kv/consul/client.go:88]: Previously we could see individual `CAS` calls to consul. Now we will only see "CAS loop".

**Comment by pracucci** [comment]
> Better way of handling multiple components with same metrics is to wrap registrerer to add constant label to all metrics registered by component

Agree. It's more powerful.

File: pkg/distributor/distributor.go [hunk 6]
```diff
@@ -184,7 +184,7 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 	replicationFactor.Set(float64(ingestersRing.ReplicationFactor()))
 	cfg.PoolConfig.RemoteTimeout = cfg.RemoteTimeout
 
-	replicas, err := newClusterTracker(cfg.HATrackerConfig)
+	replicas, err := newClusterTracker(cfg.HATrackerConfig, nil)
```
Comment (pracucci) [inline pkg/distributor/distributor.go:187]: Why don't we pass the registerer here?

File: pkg/distributor/ha_tracker.go [hunk 7]
```diff
@@ -157,7 +157,7 @@ func newClusterTracker(cfg HATrackerConfig) (*haTracker, error) {
 	}
 
 	if cfg.EnableHATracker {
-		client, err := kv.NewClient(cfg.KVStore, GetReplicaDescCodec())
+		client, err := kv.NewClient(RingNameForClient, cfg.KVStore, GetReplicaDescCodec(), reg)
```
Comment (pracucci) [inline pkg/distributor/ha_tracker.go:160]: I think using the same name here would clash with the distributor ring name. We could just pass `distributor-hatracker` as ring client name. What do you think? /cc @pstibrany

File: pkg/ring/kv/metrics.go [hunk 8]
```diff
@@ -32,12 +22,36 @@ func errorCode(err error) string {
...
-	c Client
+	c               Client
+	requestDuration *instrument.HistogramCollector
+}
+
+func newMetricsClient(name string, backend string, c Client, reg prometheus.Registerer) Client {
+	// TODO: Ensure all clients are passed a valid registry and remove the following check
+	if reg == nil {
+		reg = prometheus.DefaultRegisterer
+	}
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/ring/kv/metrics.go:33]: Agree. This TODO should be removed (and the default register never used).

File: pkg/querier/blocks_store_queryable.go [hunk 9]
```diff
@@ -139,7 +139,7 @@ func NewBlocksStoreQueryableFromConfig(querierCfg Config, gatewayCfg storegatewa
 
 	if gatewayCfg.ShardingEnabled {
 		storesRingCfg := gatewayCfg.ShardingRing.ToRingConfig()
-		storesRingBackend, err := kv.NewClient(storesRingCfg.KVStore, ring.GetCodec())
+		storesRingBackend, err := kv.NewClient(storegateway.RingNameForClient+"-block-store", storesRingCfg.KVStore, ring.GetCodec(), reg)
```
Comment (pracucci) [inline pkg/querier/blocks_store_queryable.go:142]: The generated name here looks a bit weird to me. Maybe `querier-store-gateway` may be more clear?

File: pkg/ring/kv/consul/client.go [hunk 10]
```diff
@@ -99,21 +98,17 @@ func (c *Client) Put(ctx context.Context, key string, value interface{}) error {
 		return err
 	}
 
-	return instrument.CollectedRequest(ctx, "Put", consulRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
```
Comment (pracucci) [inline pkg/ring/kv/consul/client.go:102]: Now that I notice, I think previous version was even wrong. We did track "Put" twice for 1 Put call. Once here, and once in `consul/metrics.go`. Am I missing anything?

File: pkg/ring/kv/metrics.go [hunk 11]
```diff
@@ -32,12 +22,36 @@ func errorCode(err error) string {
...
+		requestDuration: instrument.NewHistogramCollector(
+			promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+				Namespace: "cortex",
+				Name:      "kv_request_duration_seconds",
+				Help:      "Time spent on kv store requests.",
+				Buckets:   prometheus.DefBuckets,
+				ConstLabels: prometheus.Labels{
+					"name": name,
+					"type": backend,
+				},
```
Comment (pracucci) [inline pkg/ring/kv/metrics.go:46]: Please mention this change in the CHANGELOG. You should also check if it has an impact on the cortex mixin's dashboards (depends on how we group series, I haven't checked).

File: pkg/ring/lifecycler.go (hunk 3 above)
Comment (pracucci) [reply pkg/ring/lifecycler.go:151]: But maybe in another PR?

File: pkg/ring/kv/consul/metrics.go [hunk 12]
```diff
@@ -1,82 +0,0 @@
...
-
-import (
-	"context"
-
-	consul "github.com/hashicorp/consul/api"
-	"github.com/prometheus/client_golang/prometheus"
-	"github.com/weaveworks/common/instrument"
-)
-
-var consulRequestDuration = instrument.NewHistogramCollector(prometheus.NewHistogramVec(prometheus.HistogramOpts{
```
Comment (pracucci) [inline pkg/ring/kv/consul/metrics.go:11]: Please remember to mention in the CHANGELOG that the metric `cortex_consul_request_duration_seconds` has been removed.

File: pkg/ring/kv/client.go (hunk 2 above)
Comment (jtlisi) [reply pkg/ring/kv/client.go:154]: 👍 Make sense

File: pkg/ring/lifecycler.go (hunk 3 above)
Comment (jtlisi) [reply pkg/ring/lifecycler.go:151]: I will pass the Registerer but keep the lifecycler metrics global for now.

File: pkg/distributor/ha_tracker_test.go (hunk 4 above)
Comment (jtlisi) [reply pkg/distributor/ha_tracker_test.go:139]: 👍

File: pkg/ring/kv/client.go (hunk 2 above)
Comment (jtlisi) [reply pkg/ring/kv/client.go:154]: Makes sense

File: pkg/distributor/ha_tracker.go (hunk 7 above)
Comment (jtlisi) [reply pkg/distributor/ha_tracker.go]: Makes sense

File: pkg/ring/kv/consul/client.go (hunk 10 above)
Comment (jtlisi) [reply pkg/ring/kv/consul/client.go:102]: After investigating further, the consul metrics probably should remain as is. They provide lower level metrics for the client interacting with consul itself: https://github.com/hashicorp/consul/blob/master/api/kv.go

In our KV implementation CAS is retried multiple times and by using a generic metrics client we are missing out on instrumenting some of these calls. One option that may be used to solve this problem is moving retries to there own KV implementation and wrap clients with it.

Comment (jtlisi) [reply pkg/ring/kv/consul/client.go]: I am going to re-add the separate consul instrumentation and will add an issue to add a generic retry implementation for the KV store interface. WDYT?

File: pkg/querier/blocks_store_queryable.go (hunk 9 above)
Comment (jtlisi) [reply pkg/querier/blocks_store_queryable.go]: Sounds good, I wasn't sure how best to name some of these clients.

File: pkg/ring/kv/consul/metrics.go (hunk 12 above)
Comment (jtlisi) [reply pkg/ring/kv/consul/metrics.go]: I re-added this metric since it fits a different use case.

File: pkg/distributor/distributor.go (hunk 6 above)
Comment (jtlisi) [reply pkg/distributor/distributor.go]: The distributor didn't pass a registerer. I added one.

**Comment by jtlisi** [comment]
@pstibrany & @pracucci 

I updated this PR per your comments. Changes include:

- Keeping the `cortex_consul_request_duration_seconds` metric and consul instrumentation w/ note about a possible way of removing it in a future PR
- Registerers are now plumbed through to each call of the kv.NewClient
- Updated ChangeLog

I will test this PR w/ the cortex-mixin this week and ensure it operates correctly with the new metrics.

If you have anymore feedback about the naming of KV clients let me know. I wasn't sure if some of the names/suffixes I chose made sense.

File: pkg/distributor/distributor_test.go [hunk 13]
```diff
@@ -369,7 +369,7 @@ func TestDistributor_PushHAInstances(t *testing.T) {
 						KVStore:         kv.Config{Mock: mock},
 						UpdateTimeout:   100 * time.Millisecond,
 						FailoverTimeout: time.Second,
-					})
+					}, prometheus.NewRegistry())
```
Comment (pstibrany) [inline pkg/distributor/distributor_test.go:372]:
```suggestion
					}, nil)
```

Since we don't verify metrics.

File: pkg/ring/kv/metrics.go [hunk 14]
```diff
@@ -32,12 +37,35 @@ func errorCode(err error) string {
...
-	c Client
+	c               Client
+	requestDuration *instrument.HistogramCollector
+}
+
+func newMetricsClient(backend string, c Client, reg prometheus.Registerer) Client {
+	// If no Registerer is provided return the raw client
+	if reg == nil {
+		return c
+	}
```
Comment (pstibrany) [inline pkg/ring/kv/metrics.go:48]: Nit: I'd move this check into `createClient` function.

File: pkg/ring/kv/metrics.go [hunk 15]
```diff
@@ -5,19 +5,24 @@ import (
...
-func init() {
-	requestDuration.Register()
+// RegistererWithKVName wraps the provided Registerer with the KV name label. If a nil reg
+// is provided, a nil registry is returned
+func RegistererWithKVName(reg prometheus.Registerer, name string) prometheus.Registerer {
+	if reg == nil {
+		return nil
+	}
+
+	return prometheus.WrapRegistererWith(prometheus.Labels{"name": name}, reg)
```
Comment (pstibrany) [inline pkg/ring/kv/metrics.go:25]: Is it worth using something more unique, e.g. `kv-name` here?

File: pkg/ring/kv/metrics.go [hunk 16]
```diff
@@ -5,19 +5,24 @@ import (
...
 
-var requestDuration = instrument.NewHistogramCollector(prometheus.NewHistogramVec(prometheus.HistogramOpts{
-	Namespace: "cortex",
-	Name:      "kv_request_duration_seconds",
-	Help:      "Time spent on consul requests.",
-	Buckets:   prometheus.DefBuckets,
-}, []string{"operation", "status_code"}))
+var (
+	primaryLabel   = prometheus.Labels{"role": "primary"}
+	secondaryLabel = prometheus.Labels{"role": "secondary"}
```
Comment (pstibrany) [inline pkg/ring/kv/metrics.go:15]: Only used from client.go, I would move it into that file.

File: CHANGELOG.md [hunk 17]
```diff
@@ -2,6 +2,7 @@
 
 ## master / unreleased
 
+* [CHANGE] Metric `cortex_kv_request_duration_seconds` now includes `name` label to denote which client is being used as well as the `backend` label to denote the KV backend implemnetation in use. #2648
```
Comment (pstibrany) [inline CHANGELOG.md:5]:
```suggestion
* [CHANGE] Metric `cortex_kv_request_duration_seconds` now includes `name` label to denote which client is being used as well as the `backend` label to denote the KV backend implementation in use. #2648
```

File: pkg/distributor/distributor_ring.go [hunk 18]
```diff
@@ -13,6 +13,10 @@ import (
 	"github.com/cortexproject/cortex/pkg/util/flagext"
 )
 
+var (
+	RingNameForClient = "distributor"
+)
+
```
Comment (pstibrany) [inline pkg/distributor/distributor_ring.go:19]:
```suggestion
const (
	ringNameForClient = "distributor"
)
```

And move to `distributor_ring.go` where it is used.

(Alternatively, just inline it)

File: pkg/ruler/ruler.go [hunk 19]
```diff
@@ -190,7 +190,11 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable promStorage.Queryable
 	}
 
 	if cfg.EnableSharding {
-		ringStore, err := kv.NewClient(cfg.Ring.KVStore, ring.GetCodec())
+		ringStore, err := kv.NewClient(
+			cfg.Ring.KVStore,
+			ring.GetCodec(),
+			kv.RegistererWithKVName(reg, ring.RulerRingKey),
```
Comment (pstibrany) [inline pkg/ruler/ruler.go:196]:
```suggestion
			kv.RegistererWithKVName(reg, "ruler"),
```

`ring.RulerRingKey` is actually `ring`.

File: pkg/storegateway/gateway.go [hunk 20]
```diff
@@ -81,7 +81,11 @@ func NewStoreGateway(gatewayCfg Config, storageCfg cortex_tsdb.Config, logLevel
 	}
 
 	if gatewayCfg.ShardingEnabled {
-		ringStore, err = kv.NewClient(gatewayCfg.ShardingRing.KVStore, ring.GetCodec())
+		ringStore, err = kv.NewClient(
+			gatewayCfg.ShardingRing.KVStore,
+			ring.GetCodec(),
+			kv.RegistererWithKVName(reg, RingNameForClient),
```
Comment (pstibrany) [inline pkg/storegateway/gateway.go:87]:
```suggestion
			kv.RegistererWithKVName(reg, "store-gateway"),
```

File: pkg/ring/kv/metrics.go (hunk 15 above)
Comment (jtlisi) [reply pkg/ring/kv/metrics.go:25]: That makes sense. I was just basing it off the label we use for the ring.

**Pull request merged**